### PR TITLE
More routing tests and bug fix

### PR DESCRIPTION
**Pull request by conorbros**
- Add routing tests for compound and composite primary keys.
- Fix a bug in the routing key serialising
- Add tests against cdrs-tokio aswell

### CODE REVIEW

File: shotover-proxy/tests/cassandra_int_tests/cluster/single_rack_v4.rs [hunk 1]
```diff
@@ -63,7 +63,7 @@ async fn test_rewrite_system_peers_dummy_peers(connection: &CassandraConnection)
         ResultValue::Any,
         // Unfortunately token generation appears to be non-deterministic but we can at least assert that
         // there are 128 tokens per node
-        ResultValue::Set(std::iter::repeat(ResultValue::Any).take(3 * 128).collect()),
+        ResultValue::AnyList(std::iter::repeat(ResultValue::Any).take(3 * 128).collect()),
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/cluster/single_rack_v4.rs:66]: I dont like that this makes tests against all drivers weaker when its only one driver that doesnt support sets properly.
Is there a reason our old workaround is no longer suitable?

File: shotover-proxy/tests/cassandra_int_tests/mod.rs [hunk 2]
```diff
@@ -145,7 +146,9 @@ async fn cluster_single_rack_v3(#[case] driver: CassandraDriver) {
         };
         standard_test_suite(&connection, driver).await;
         cluster::single_rack_v3::test_dummy_peers(&connection().await).await;
-        routing::test("127.0.0.1", 9042, "172.16.1.2", 9042).await;
+
+        #[cfg(not(feature = "cassandra-cpp-driver-tests"))]
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:150]: why?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs [hunk 3]
```diff
@@ -54,7 +56,7 @@ fn serialize_routing_key_with_indexes(
 }
 
 // https://github.com/apache/cassandra/blob/3a950b45c321e051a9744721408760c568c05617/src/java/org/apache/cassandra/db/marshal/CompositeType.java#L39
-fn serialize_routing_value(cursor: &mut Cursor<&mut Vec<u8>>, value: &Value, version: Version) {
+fn serialize_routing_value(cursor: &mut Cursor<&mut Vec<u8>>, value: &Vec<u8>, version: Version) {
     let temp_size: CIntShort = 0;
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs:60]: We can now change this to `let size: CIntShort = value.len()`, allowing us to remove the:
```rust
    let after_value_pos = cursor.position();
    cursor.set_position(before_value_pos - SHORT_LEN as u64);
    let value_size: CIntShort = (after_value_pos - before_value_pos) as CIntShort;
    value_size.serialize(cursor, version);
```

File: shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs [hunk 4]
```diff
@@ -54,7 +56,7 @@ fn serialize_routing_key_with_indexes(
 }
 
 // https://github.com/apache/cassandra/blob/3a950b45c321e051a9744721408760c568c05617/src/java/org/apache/cassandra/db/marshal/CompositeType.java#L39
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs:58]: The linked source mentions the final byte being one of -1, 0 or 1, but we just hardcode to 0.
I dont really understand whats going on here but do we need to use any of those other values in certain cases?

File: shotover-proxy/tests/cassandra_int_tests/mod.rs (hunk 2 above)
Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/mod.rs:150]: Don't have tracing support on cassandra-cpp

File: shotover-proxy/tests/cassandra_int_tests/cluster/single_rack_v4.rs (hunk 1 above)
Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/cluster/single_rack_v4.rs:66]: What was the old work around?

Comment (rukai) [reply shotover-proxy/tests/cassandra_int_tests/cluster/single_rack_v4.rs:66]: https://github.com/shotover/shotover-proxy/blob/5a1558bc0ca35e63ec614569f2540db11aa33e2c/shotover-proxy/tests/cassandra_int_tests/collections.rs#L194

File: shotover-proxy/tests/cassandra_int_tests/mod.rs (hunk 2 above)
Comment (rukai) [reply shotover-proxy/tests/cassandra_int_tests/mod.rs:150]: Oh I see, previously we were completely ignoring the driver in routing::test and thats why it used to work. :+1:

File: shotover-proxy/tests/cassandra_int_tests/routing.rs [hunk 5]
```diff
@@ -1,93 +1,336 @@
...
+        run_query(connection, create_ks).await;
+    }
+
+    async fn create_table(connection: &mut CassandraConnection) {
+        let create_table_cql =
+        "CREATE TABLE IF NOT EXISTS test_routing_ks.my_test_table_composite (key int, name text, age int, blah text, PRIMARY KEY((key, name), age));";
+        run_query(connection, create_table_cql).await;
+    }
+
+    pub async fn test(shotover: &mut CassandraConnection, cassandra: &CassandraConnection) {
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/routing.rs:199]: the `shotover` connection doesnt need to be `&mut` as demonstrated by the `cassandra` connection which is just `&`

File: shotover-proxy/tests/cassandra_int_tests/mod.rs [hunk 6]
```diff
@@ -155,7 +158,7 @@ async fn cluster_single_rack_v3(#[case] driver: CassandraDriver) {
 }
 
 #[rstest]
-//#[case::cdrs(CdrsTokio)] // TODO
+#[case::cdrs(CdrsTokio)] // TODO
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:161]: TODO got left here

**Review by rukai — APPROVED** [review]
Nice!
Appreciating the increase coverage on the integration tests.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs (hunk 4 above)
Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs:57]: https://github.com/scylladb/scylladb/blob/40adf38915b6d8f5314c621a94d694d172360833/compound_compat.hh#L46

https://github.dev/datastax/java-driver/blob/a650ee43712cc0448ce6d263db30f15eb2f55c2c/core/src/main/java/com/datastax/oss/driver/internal/core/util/RoutingKey.java#L36

Some examples of other drivers doing the same.

Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/routing_key.rs:57]: The first one is scylladb itself, not a driver, right?

If the java driver just hardcodes to 0, then it must be something fancy the server supports but the drivers can get away with ignoring.
Thanks for investigating :+1:

**Review by XA21X — APPROVED** [review]

**Pull request merged**